crates/shared/src/transport/buffered.rs

```diff
@@ -0,0 +1,305 @@
+    web3::{BatchTransport, Error as Web3Error, RequestId, Transport},
+};
+use futures::{
+    channel::{mpsc, oneshot},
+    future::{self, BoxFuture, FutureExt as _},
+    stream::{self, FusedStream, Stream, StreamExt as _},
+};
+use serde_json::Value;
+use std::{future::Future, sync::Arc, time::Duration};
+use tokio::task::JoinHandle;
+
+/// Buffered transport configuration.
+pub struct Configuration {
+    /// The maximum amount of concurrent batches to send to the node.
+    pub max_concurrent_requests: usize,
+    /// The maximum batch size.
+    pub max_batch_len: usize,
+    /// An additional minimum delay to wait for collecting requests.
+    ///
+    /// The delay starts counting after receiving the first request.
+    pub batch_delay: Duration,
+}
+
+impl Default for Configuration {
+    fn default() -> Self {
+        // Default configuration behaves kind of like TCP Nagle.
+        Self {
+            max_concurrent_requests: 1,
+            max_batch_len: MAX_BATCH_SIZE,
+            batch_delay: Duration::default(),
+        }
+    }
+}
+
+/// Buffered `Transport` implementation that implements automatic batching of
+/// JSONRPC requests.
+#[derive(Clone, Debug)]
+pub struct Buffered<Inner> {
+    inner: Arc<Inner>,
+    calls: mpsc::UnboundedSender<CallContext>,
+}
+
+type RpcResult = Result<Value, Web3Error>;
+
+type CallContext = (RequestId, Call, oneshot::Sender<RpcResult>);
+
+impl<Inner> Buffered<Inner>
+where
+    Inner: BatchTransport + Send + Sync + 'static,
+    Inner::Out: Send,
+    Inner::Batch: Send,
+{
+    /// Create a new buffered transport with the default configuration.
+    pub fn new(inner: Inner) -> Self {
+        Self::with_config(inner, Default::default())
+    }
+
+    /// Creates a new buffered transport with the specified configuration.
+    pub fn with_config(inner: Inner, config: Configuration) -> Self {
+        let inner = Arc::new(inner);
+        let (calls, receiver) = mpsc::unbounded();
+        Self::background_worker(inner.clone(), config, receiver);
+
+        Self { inner, calls }
+    }
+
+    /// Start a background worker for handling batched requests.
+    fn background_worker(
+        inner: Arc<Inner>,
+        config: Configuration,
+        calls: mpsc::UnboundedReceiver<CallContext>,
+    ) -> JoinHandle<()> {
+        tokio::task::spawn(batched_for_each(config, calls, move |mut batch| {
+            let inner = inner.clone();
+            async move {
+                match batch.len() {
+                    0 => (),
+                    1 => {
+                        let (id, request, sender) = batch.remove(0);
+                        let result = inner.send(id, request).await;
+                        let _ = sender.send(result);
+                    }
+                    n => {
+                        let (requests, senders): (Vec<_>, Vec<_>) = batch
+                            .into_iter()
+                            .filter(|(_, _, sender)| sender.is_canceled())
```

Comment:
Missing NOT ?

Reply:
:see_no_evil: I was wondering why my tests were failing.